### PR TITLE
Particle Container to Pure SoA Again

### PR DESCRIPTION
**Pull request by ax3l**
Transition to new, purely SoA particle containers.

This was originally merged in #3850 and reverted in #4652, since we discovered issues loosing particles & laser particles on GPU.

- [x] update AMReX again with Jan 30/31 changes #4654
- [x] improve performance of id handling via constants and https://github.com/AMReX-Codes/amrex/pull/3735
- [x] update docs to make sure AoS is not mentioned anymore and particle id handling is guided
- [x] run test suite on CUDA devices with high checksum tolerance

## Fun Mini-Benchmarks on CPU, DP

Hardware: 12th Gen Intel(R) Core(TM) i9-12900H

```bash
export OMP_NUM_THREADS=1
./warpx.3d ../../Examples/Tests/performance_tests/automated_test_1_uniform_rest_32ppc amr.max_grid_size=64 amr.n_cell=64 64 64 max_step=5 &
taskset -cp 6 $!
```

[cpu_legacy.txt](https://github.com/ECP-WarpX/WarpX/files/14149010/cpu_legacy.txt), [cpu_soa.txt](https://github.com/ECP-WarpX/WarpX/files/14149011/cpu_soa.txt)

Overall speed: similar to noise level of repeated runs (as expected).

Few noteworthy details in top 10 functions by runtime (Excl.):
- `ParticleContainer::RedistributeCPU`: 8% slower :eyes: -> https://github.com/AMReX-Codes/amrex/issues/3744
- `WarpXParticleContainer::ApplyBoundaryConditions`: 5% faster
- `WarpX::OneStep_nosub`: 2% slower :eyes: 

## Fun Mini-Benchmarks on A100 GPU, DP

Hardware: Perlmutter (NERSC) A100 GPU

```bash
./warpx.3d ../../Examples/Tests/performance_tests/automated_test_1_uniform_rest_32ppc amr.max_grid_size=256 amr.n_cell=256 256 256 max_step=5
```

Overall speed: 1.4% faster

Few noteworthy details in top 10 functions by runtime (Excl.):
- `GatherAndPush`: 1.2% faster
- `Redistribute_partition`: 4% faster
- `AddPlasma`: 2.6% faster
- `ApplyBoundaryConditions`: 1% faster
- `SortParticlesForDeposition`: 231% faster :rocket: :rocket: :sparkles:
- `PermutationForDeposition`: 3% faster
- `InitData`: 15% faster :rocket:
- rest in TOP10 is the same as before

### CODE REVIEW

File: Source/Diagnostics/ReducedDiags/FieldProbeParticleContainer.cpp [hunk 1]
```diff
@@ -89,33 +89,17 @@ FieldProbeParticleContainer::AddNParticles (int lev,
...
-        amrex::ignore_unused(x, y);
-        p.pos(0) = z[i];
-#endif
-
-        // write position, cpu id, and particle id to particle
-        pinned_tile.push_back(p);
+        auto & idcpu_data = pinned_tile.GetStructOfArrays().GetIdCPUData();
+        idcpu_data.push_back(0);
+        amrex::ParticleIDWrapper{idcpu_data.back()} = ParticleType::NextID();
+        amrex::ParticleCPUWrapper(idcpu_data.back()) = ParallelDescriptor::MyProc();
```
Comment (ax3l) [inline Source/Diagnostics/ReducedDiags/FieldProbeParticleContainer.cpp]: Let's use https://github.com/AMReX-Codes/amrex/pull/3733

File: Source/Particles/ParticleCreation/SmartCreate.H [hunk 2]
```diff
@@ -47,23 +49,23 @@ struct SmartCreate
...
 #else
-        prt.m_aos[i_prt].pos(0) = z;
+        prt.m_rdata[PIdx::z][i_prt] = z;
         amrex::ignore_unused(x,y);
 #endif
 
-        prt.m_aos[i_prt].cpu() = cpu;
-        prt.m_aos[i_prt].id() = id;
+        amrex::ParticleIDWrapper{prt.m_idcpu[i_prt]} = id;
+        amrex::ParticleCPUWrapper{prt.m_idcpu[i_prt]} = cpu;
```
Comment (ax3l) [inline Source/Particles/ParticleCreation/SmartCreate.H]: Let's use Let's use https://github.com/AMReX-Codes/amrex/pull/3733

File: Source/Particles/ParticleCreation/SmartUtils.H [hunk 3]
```diff
@@ -60,12 +61,12 @@ void setNewParticleIDs (PTile& ptile, int old_size, int num_added)
...
-    auto pp = ptile.GetArrayOfStructs()().data() + old_size;
+    auto ptd = ptile.getParticleTileData();
     amrex::ParallelFor(num_added, [=] AMREX_GPU_DEVICE (int ip) noexcept
     {
-        auto& p = pp[ip];
-        p.id() = pid+ip;
-        p.cpu() = cpuid;
+        auto const new_id = ip + old_size;
+        amrex::ParticleIDWrapper{ptd.m_idcpu[new_id]} = pid+ip;
+        amrex::ParticleCPUWrapper{ptd.m_idcpu[new_id]} = cpuid;
```
Comment (ax3l) [inline Source/Particles/ParticleCreation/SmartUtils.H]: Let's use Let's use https://github.com/AMReX-Codes/amrex/pull/3733

File: Source/Particles/WarpXParticleContainer.cpp [hunk 4]
```diff
@@ -198,52 +198,54 @@ WarpXParticleContainer::AddNParticles (int /*lev*/, long n,
...
+        auto & idcpu_data = pinned_tile.GetStructOfArrays().GetIdCPUData();
+        idcpu_data.push_back(0);
+        if (id==-1) {
+            amrex::ParticleIDWrapper{idcpu_data.back()} = ParticleType::NextID();
         } else {
-            p.id() = id;
+            amrex::ParticleIDWrapper{idcpu_data.back()} = id;
         }
-        p.cpu() = amrex::ParallelDescriptor::MyProc();
+        amrex::ParticleCPUWrapper(idcpu_data.back()) = ParallelDescriptor::MyProc();
```
Comment (ax3l) [inline Source/Particles/WarpXParticleContainer.cpp]: Let's use https://github.com/AMReX-Codes/amrex/pull/3733

File: Source/Particles/Collision/BinaryCollision/DSMC/SplitAndScatterFunc.H [hunk 5]
```diff
@@ -70,20 +71,37 @@ int splitScatteringParticles (
...
+                constexpr amrex::Long minus_one_long = -1;
+                uint64_t tmp = 0;
+                amrex::ParticleIDWrapper wrapper(tmp);
+                wrapper = minus_one_long;
+#if defined(AMREX_USE_OMP)
+#pragma omp atomic write
+                idcpu = wrapper.m_idata;
+#else
+                auto *old_ptr = reinterpret_cast<unsigned long long*>(&idcpu);
+                    amrex::Gpu::Atomic::Exch(old_ptr, (unsigned long long) wrapper.m_idata);
```
Comment (ax3l) [inline Source/Particles/Collision/BinaryCollision/DSMC/SplitAndScatterFunc.H]:
```suggestion
#if defined(AMREX_USE_OMP)
#pragma omp atomic write
                idcpu = amrex::ParticleIdCpus::Invalid;
#else
                auto *old_ptr = reinterpret_cast<unsigned long long*>(&idcpu);
                    amrex::Gpu::Atomic::Exch(old_ptr, amrex::ParticleIdCpus::Invalid);
```

File: Source/Particles/Collision/BinaryCollision/ParticleCreationFunc.H [hunk 6]
```diff
@@ -205,16 +202,32 @@ public:
...
+                    constexpr amrex::Long minus_one_long = -1;
+                    uint64_t tmp = 0;
+                    amrex::ParticleIDWrapper wrapper(tmp);
+                    wrapper = minus_one_long;
+#if defined(AMREX_USE_OMP)
+#pragma omp atomic write
+                    idcpu = wrapper.m_idata;
+#else
+                    auto *old_ptr = reinterpret_cast<unsigned long long*>(&idcpu);
+                    amrex::Gpu::Atomic::Exch(old_ptr, (unsigned long long) wrapper.m_idata);
```
Comment (ax3l) [inline Source/Particles/Collision/BinaryCollision/ParticleCreationFunc.H]:
```suggestion
#if defined(AMREX_USE_OMP)
#pragma omp atomic write
                    idcpu = amrex::ParticleIdCpus::Invalid;
#else
                    auto *old_ptr = reinterpret_cast<unsigned long long*>(&idcpu);
                    amrex::Gpu::Atomic::Exch(old_ptr, amrex::ParticleIdCpus::Invalid);
```

File: Source/Particles/ElementaryProcess/QEDPairGeneration.H [hunk 7]
```diff
@@ -167,7 +167,7 @@ public:
             p_ux, p_uy, p_uz,
             engine);
 
-        src.m_aos[i_src].id() = -1; //destroy photon after pair generation
+        amrex::ParticleIDWrapper{src.m_idcpu[i_src]} = -1;  // destroy photon after pair generation
```
Comment (ax3l) [inline Source/Particles/ElementaryProcess/QEDPairGeneration.H]:
```suggestion
        src.m_idcpu[i_src] = amrex::ParticleIdCpus::Invalid;  // destroy photon after pair generation
```

File: Source/Particles/ElementaryProcess/QEDPhotonEmission.H [hunk 8]
```diff
@@ -262,8 +260,8 @@ void cleanLowEnergyPhotons(
         constexpr amrex::ParticleReal me_c = PhysConst::m_e*PhysConst::c;
         const auto phot_energy2 = (ux*ux + uy*uy + uz*uz)*me_c*me_c;
 
-        if (phot_energy2 < energy_threshold2){
-            p.id() = - 1;
+        if (phot_energy2 < energy_threshold2) {
+            amrex::ParticleIDWrapper{p_idcpu[ip]} = -1;
```
Comment (ax3l) [inline Source/Particles/ElementaryProcess/QEDPhotonEmission.H]:
```suggestion
            p_idcpu[ip] = amrex::ParticleIdCpus::Invalid;
```

File: Source/Particles/Resampling/LevelingThinning.cpp [hunk 9]
```diff
@@ -114,7 +113,7 @@ void LevelingThinning::operator() (WarpXParIter& pti, const int lev,
                 // Remove particle with probability 1 - particle_weight/level_weight
                 if (random_number > w[indices[i]]/level_weight)
                 {
-                    particle_ptr[indices[i]].id() = -1;
+                    amrex::ParticleIDWrapper{idcpu[indices[i]]} = -1;
```
Comment (ax3l) [inline Source/Particles/Resampling/LevelingThinning.cpp]:
```suggestion
                    idcpu[indices[i]] = amrex::ParticleIdCpus::Invalid;
```

File: Source/Particles/WarpXParticleContainer.cpp [hunk 10]
```diff
@@ -1504,7 +1503,7 @@ WarpXParticleContainer::ApplyBoundaryConditions (){
                                                               boundary_conditions, engine);
 
                     if (particle_lost) {
-                        p.id() = -p.id();
+                        amrex::ParticleIDWrapper{idcpu[i]} = -id;
```
Comment (ax3l) [inline Source/Particles/WarpXParticleContainer.cpp]: Let's use https://github.com/AMReX-Codes/amrex/pull/3735

**Comment by ax3l** [comment]
memo to self:
- GPU test in screen from PM `login40`

**Comment by ax3l** [comment]
# GPU Tests (CUDA, A100 on Perlmutter)

```diff
diff --git a/Examples/analysis_default_openpmd_regression.py b/Examples/analysis_default_openpmd_regression.py
index 3aadc49ac5..3e9fb98789 100755
--- a/Examples/analysis_default_openpmd_regression.py
+++ b/Examples/analysis_default_openpmd_regression.py
@@ -15,6 +15,6 @@ test_name = os.path.split(os.getcwd())[1]
 
 # Run checksum regression test
 if re.search( 'single_precision', fn ):
-    checksumAPI.evaluate_checksum(test_name, fn, output_format='openpmd', rtol=2.e-6)
+    checksumAPI.evaluate_checksum(test_name, fn, output_format='openpmd', rtol=4.)
 else:
-    checksumAPI.evaluate_checksum(test_name, fn, output_format='openpmd')
+    checksumAPI.evaluate_checksum(test_name, fn, output_format='openpmd', rtol=4.)
diff --git a/Examples/analysis_default_regression.py b/Examples/analysis_default_regression.py
index 453f650be0..6fa855df3d 100755
--- a/Examples/analysis_default_regression.py
+++ b/Examples/analysis_default_regression.py
@@ -15,6 +15,6 @@ test_name = os.path.split(os.getcwd())[1]
 
 # Run checksum regression test
 if re.search( 'single_precision', fn ):
-    checksumAPI.evaluate_checksum(test_name, fn, rtol=2.e-6)
+    checksumAPI.evaluate_checksum(test_name, fn, rtol=4.)
 else:
-    checksumAPI.evaluate_checksum(test_name, fn)
+    checksumAPI.evaluate_checksum(test_name, fn, rtol=4.)
diff --git a/Regression/WarpX-tests.ini b/Regression/WarpX-tests.ini
index 3310e642dd..84133add09 100644
--- a/Regression/WarpX-tests.ini
+++ b/Regression/WarpX-tests.ini
@@ -40,7 +40,7 @@ use_ctools = 0
 # sections.
 
 #MPIcommand = mpiexec -host @host@ -n @nprocs@ @command@
-MPIcommand = mpiexec -n @nprocs@ @command@
+MPIcommand = srun -n @nprocs@ @command@
 MPIhost =
 
 reportActiveTestsOnly = 1
@@ -64,7 +64,7 @@ branch = 24.02
 [source]
 dir = /home/regtester/AMReX_RegTesting/warpx
 branch = development
-cmakeSetupOpts = -DAMReX_ASSERTIONS=ON -DAMReX_TESTING=ON -DWarpX_PYTHON_IPO=OFF -DpyAMReX_IPO=OFF
+cmakeSetupOpts = -DAMReX_ASSERTIONS=ON -DAMReX_TESTING=ON -DWarpX_PYTHON_IPO=OFF -DpyAMReX_IPO=OFF -DWarpX_COMPUTE=CUDA
 # -DPYINSTALLOPTIONS="--disable-pip-version-check"
 
 # individual problems follow
```
```bash
cat test.sbatch 
```
```
#!/bin/bash -l

# Copyright 2021-2023 Axel Huebl, Kevin Gott
#
# This file is part of WarpX.
#
# License: BSD-3-Clause-LBNL

#SBATCH -t 10:00:00
#SBATCH -N 1
#SBATCH -J run_test_soa
#SBATCH -A m4272_g
#SBATCH -q regular
# A100 40GB (most nodes)
#SBATCH -C gpu
# A100 80GB (256 nodes)
#S BATCH -C gpu&hbm80g
#SBATCH --exclusive
# ideally single:1, but NERSC cgroups issue
#SBATCH --gpu-bind=none
#SBATCH --ntasks-per-node=4
#SBATCH --gpus-per-node=4
#SBATCH -o WarpX.o%j
#SBATCH -e WarpX.e%j

# pin to closest NIC to GPU
export MPICH_OFI_NIC_POLICY=GPU

# threads for OpenMP and threaded compressors per MPI rank
#   note: 16 avoids hyperthreading (32 virtual cores, 16 physical)
export SRUN_CPUS_PER_TASK=16

export WARPX_CI_NUM_MAKE_JOBS=32

./run_test.sh
```

## Tests that pass within a 10hr walltime in `development`

53 (until walltime reached)

## Tests that pass within a 10hr walltime with this PR

53 (until walltime reached)

## Tests that Already Crash in `development`

- ImplicitPicard_VandB_2d CRASHED (backtraces produced)
- Langmuir_multi_2d_MR CRASHED (backtraces produced)
- Langmuir_multi_2d_MR_momentum_conserving CRASHED (backtraces produced)
- Langmuir_multi_2d_MR_psatd CRASHED (backtraces produced)
- LaserInjection CRASHED (backtraces produced)
- ...

Some of those crash because of our default warning threshold, so I can retry with those.
```
!!! WARNING : [high][Performance] Too many boxes per GPU!
...
1::Assertion `msg_priority < abort_priority' failed, file "/tmp/ci-bbDk7v5Td5/warpx/Source/ablastr/warn_manager/WarnManager.cpp", line 97, Msg: 
```

## Tests that Crash with this PR

- ImplicitPicard_VandB_2d CRASHED (backtraces produced)
- Langmuir_multi_2d_MR CRASHED (backtraces produced)
- Langmuir_multi_2d_MR_momentum_conserving CRASHED (backtraces produced)
- Langmuir_multi_2d_MR_psatd CRASHED (backtraces produced)
- LaserInjection CRASHED (backtraces produced)


## `rtol=4.` Checksums

### Tests that Already Fail Analysis in `development`

- BTD_rz FAILED
- Deuterium_Deuterium_Fusion_3D FAILED
- Deuterium_Deuterium_Fusion_3D_intraspecies FAILED
- Deuterium_Tritium_Fusion_3D FAILED
- Deuterium_Tritium_Fusion_RZ FAILED
- ElectrostaticSphereRZ FAILED
- FluxInjection FAILED
- FluxInjection3D FAILED
- ImplicitPicard_1d FAILED
- Langmuir_multi_psatd_single_precision FAILED
- Langmuir_multi_rz FAILED
- Langmuir_multi_rz_psatd FAILED
- Langmuir_multi_rz_psatd_current_correction FAILED
- Langmuir_multi_single_precision FAILED
- LaserAcceleration_BTD FAILED
- LaserInjectionFromLASYFile_RZ FAILED
...

### Tests that Already Fail with this PR

- BTD_rz FAILED
- Deuterium_Deuterium_Fusion_3D FAILED
- Deuterium_Deuterium_Fusion_3D_intraspecies FAILED
- Deuterium_Tritium_Fusion_3D FAILED
- Deuterium_Tritium_Fusion_RZ FAILED
- ElectrostaticSphereRZ FAILED
- FluxInjection FAILED
- FluxInjection3D FAILED
- ImplicitPicard_1d FAILED
- Langmuir_multi_psatd_single_precision FAILED
- Langmuir_multi_rz FAILED
- Langmuir_multi_rz_psatd FAILED
- Langmuir_multi_rz_psatd_current_correction FAILED
- Langmuir_multi_single_precision FAILED
- LaserAcceleration_BTD FAILED
- LaserInjectionFromLASYFile_RZ FAILED
...

**Comment by RemiLehe** [comment]
> Tests that Crash with this PR

File: Source/Diagnostics/WarpXOpenPMD.cpp [hunk 11]
```diff
@@ -722,77 +727,7 @@ WarpXOpenPMDPlot::DumpToFile (ParticleContainer* pc,
...
-                           [](amrex::ParticleReal const *p) { delete[] p; }
-                   );
-                   for (auto i = 0; i < numParticleOnTile; i++) {
-                       z.get()[i] = aos[i].pos(1);  // {0: "r", 1: "z"}
-                   }
-                   std::string const positionComponent = "z";
-                   currSpecies["position"]["z"].storeChunk(z, {offset}, {numParticleOnTile64});
-                }
-
-                //   reconstruct x and y from polar coordinates r, theta
```
Comment (ax3l) [inline Source/Diagnostics/WarpXOpenPMD.cpp:744]: Oopsi, reconstruction re-added in #4686

File: Source/Particles/PhysicalParticleContainer.cpp [hunk 12]
```diff
@@ -1084,7 +1083,7 @@ PhysicalParticleContainer::AddPlasma (PlasmaInjector const& plasma_injector, int
         const int max_new_particles = Scan::ExclusiveSum(counts.size(), counts.data(), offset.data());
 
         // Update NextID to include particles created in this function
-        Long pid;
+        int pid;
```
Comment (ax3l) [inline Source/Particles/PhysicalParticleContainer.cpp:1086]: Long!

File: Source/Particles/ParticleCreation/SmartUtils.H [hunk 13]
```diff
@@ -60,12 +61,11 @@ void setNewParticleIDs (PTile& ptile, int old_size, int num_added)
...
 
     const int cpuid = amrex::ParallelDescriptor::MyProc();
-    auto pp = ptile.GetArrayOfStructs()().data() + old_size;
+    auto ptd = ptile.getParticleTileData();
     amrex::ParallelFor(num_added, [=] AMREX_GPU_DEVICE (int ip) noexcept
     {
-        auto& p = pp[ip];
-        p.id() = pid+ip;
-        p.cpu() = cpuid;
+        auto const new_id = ip + old_size;
```
Comment (ax3l) [inline Source/Particles/ParticleCreation/SmartUtils.H:67]: Long!

**Pull request merged**